_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,193 @@
+
+We can notice a pattern: **it is still possible to support different use cases and configuration for a component with multiple responsibilities, but only by increasing the size and the complexity of the component itself**, like adding flags and conditional logic. Little by little, this is how [big ball of mud](https://en.wikipedia.org/wiki/Big_ball_of_mud) systems (aka monoliths), and [runaway methods](https://michaelfeathers.typepad.com/michael_feathers_blog/2012/09/runaway-methods.html) and classes emerge. When each component has a single responsibility, we can keep the complexity of any single one of them limited.
+
+What about the "lean" approach of splitting responsibilities only when we actually need to make them configurable? I think this is a good strategy, if applied with moderation. It is similar to the Martin Fowler's idea of [preparatory refactoring](https://www.martinfowler.com/articles/preparatory-refactoring-example.html). Keep in mind, however, that **if we don't keep responsibilities separate from early on, they may "fuse" into one another as the code evolves, so it might take much more effort to split them apart further down the road.** To do this, we might also need to spend time relearning the workings of the code in more detail than we would otherwise like to.
+
+### Reusability
+
+**When a component has a single, narrow responsibility, it's easier to reuse it.** `FraudDetection` class from the previous section is an example of this: we could reuse it in online processing and batch processing components. To do this in the *artifact space*, we could pull it into a shared library. Another direction is to move fraud detection into a separate microservice, which we can think about as reusability in the *runtime space*. The `FraudDetection` class within our application will then turn from having business logic responsibility to do external integration with the new service.
+
+**Most methods with a narrow responsibility shouldn't have side effects and shouldn't depend on the state of the class**, which enables sharing and calling them from any place. In other words, Single Responsibility Principle nudges us toward [functional programming](https://en.wikipedia.org/wiki/Functional_programming) style.
+
+***
+Pro tip: thinking about responsibilities helps to notice [unrelated subproblems](https://learning.oreilly.com/library/view/the-art-of/9781449318482/ch10.html) hiding in our methods and classes. When we extract them, we can then see opportunities to reuse them in other places. Moving unrelated subproblems out of the way keeps a component at the [single level or abstraction](http://principles-wiki.net/principles:single_level_of_abstraction), which makes easier to understand the logic of the component.
+***
```

Comment:
This is a candidate for an info box, I'd say. Would be good to have a couple to break up the text.

Reply:
Agreed